Urllib2

pmhahn opened the pull request:
Several more type annotation for urllib2 in Python2.
The first 4 patches should be uncontroversial.
Modeling the Response helped me find some bug where I accessed internal data.
Please have a close look at the last patch as `Message` is not a subclass of `dict`, so it might breaks old uses where using `dict` was okay. `urllib2` itself will always pass a `Message` instance, but custom code might create its own dictionary and pass it back to further calls.

srittau commented on stdlib/2/urllib2.pyi:
```diff
@@ -11,8 +13,8 @@ class URLError(IOError):
-    headers = ...  # type: Dict[str, str]
-    def __init__(self, url, code, msg, hdrs, fp) -> None: ...
+    headers = ...  # type: Message
+    def __init__(self, url, code: int, msg: str, hdrs: Message, fp: IO) -> None: ...
```
Ideally, we would use a protocol for the `fp` pointer as it's only required to have `read()` and `readline()` methods, according to the [implementation of `addbase`](https://github.com/python/cpython/blob/34b7c438b8dc0a1e7e23c9b2d7ce7f8a7c31b4f4/Lib/urllib.py#L955-L970). But for now, I'd be content with just using `Any`. Also applies to a few more functions below.

pmhahn replied:
836ca955 is my first stab at protocols; do you prefer that or should I go for `Any`?

srittau commented on stdlib/2/urllib2.pyi:
```diff
@@ -11,8 +13,8 @@ class URLError(IOError):
     code = ...  # type: int
-    headers = ...  # type: Dict[str, str]
-    def __init__(self, url, code, msg, hdrs, fp) -> None: ...
+    headers = ...  # type: Message
```
Nit: For new and changed code we prefer variable annotation syntax: `headers: Message`. (Or rather `header: Mapping[str, str]` in this case.)

pmhahn replied:
Fixed by fc1ef1bcacbeebff596a994a8831e6df80bdb07c and fc2feed713938f1a74b59917db25c792f71ad02f

srittau commented on stdlib/2/urllib2.pyi:
```diff
@@ -45,14 +47,21 @@ class Request(object):
+
     def add_handler(self, handler: BaseHandler) -> None: ...
-    def open(self, url: Union[Request, _string], data: Optional[_string] = ..., timeout: Optional[float] = ...): ...
+    def open(self, url: Union[Request, _string], data: Optional[_string] = ..., timeout: Optional[float] = ...) -> Optional[_Response]: ...
```
The first argument is called `fullurl` (important when someone uses keyword argument syntax).

The `timeout` parameter is a bit problematic. It accepts float values and defaults to the singleton `socket._GLOBAL_DEFAULT_TIMEOUT`, which is just an `object` instance. But I don't think `None` is accepted, so it should not be `Optional`.

pmhahn replied:
I fixed `OpenerDirector.open(fullurl)` in 3039be30edf94b227d3c5512fc4ef771046a87f7. Thanks for your feedback.

About `timeout`: `urlib2` delegates to `urllib` delegates to `ftplib` and `httplib`, which both use `socket`. Quoting <https://docs.python.org/2/library/socket.html#socket.socket.settimeout>:
> ... The value argument can be a nonnegative float expressing seconds, or None. ... Setting a timeout of None disables timeouts on socket operations.
And yes, passing `None` to `urllib2.urlopen(timeout)` is passed all the way down to `socket.settimeout()` and disables the timeout. That may be a gory implementation detail, but is useful:
```python
import socket, urllib2
socket.setdefaulttimeout(0.01)
c = urllib2.urlopen('https://python.org/', timeout=None)
```
See commit message of 46d856847e5bcced97b974fa9060212e9cb71674 for more details.

srittau commented on stdlib/2/urllib2.pyi:
```diff
@@ -68,50 +77,52 @@ class HTTPErrorProcessor(BaseHandler):
-    def add_password(self, realm: _string, uri: Union[_string, Sequence[_string]], user: _string, passwd: _string) -> None: ...
-    def find_user_password(self, realm: _string, authuri: _string) -> Tuple[Any, Any]: ...
+    def add_password(self, realm: Optional[_string], uri: Union[_string, Sequence[_string]], user: _string, passwd: _string) -> None: ...
+    def find_user_password(self, realm: Optional[_string], authuri: _string) -> Tuple[Any, Any]: ...
```
Is `realm` really supposed to be `None`? While the implementation would work with `None` values, lines like the following make me assume that this is not really a supported use case:

https://github.com/python/cpython/blob/34b7c438b8dc0a1e7e23c9b2d7ce7f8a7c31b4f4/Lib/urllib2.py#L787

Also applies to two more instances below.

pmhahn replied:
Quoting the commit message of my own b7744a9c872bf75688e4fe5392cfc71a8eeea0f6:
> Also realm=None is allowed, which is required when
> HTTPPasswordMgrWithDefaultRealm is used:
> <https://docs.python.org/2/howto/urllib2.html#id6>:
> > Frequently one doesn’t care what the realm is.
> ...
> > We indicate this by providing None as the realm argument to the
> > add_password method.

See <https://github.com/python/cpython/blob/34b7c438b8dc0a1e7e23c9b2d7ce7f8a7c31b4f4/Lib/urllib2.py#L849> where `find_user_password(self, realm, authuri)` is called with `realm=None`

JelleZijlstra commented on stdlib/2/urllib2.pyi:
```diff
@@ -1,18 +1,29 @@
-    headers = ...  # type: Dict[str, str]
-    def __init__(self, url, code, msg, hdrs, fp) -> None: ...
+    headers: Mapping[str, str] = ...
+    def __init__(self, url, code: int, msg: str, hdrs: Mapping[str, str], fp: _SupportsRead) -> None: ...
```
As far as I can tell this only needs read, readlines, and close, not the much bigger protocol you wrote. (I'm going off the implementation of `urllib.addbase`.)

JelleZijlstra replied:
What about `close`? IIRC the close method of addbase uses it.

pmhahn replied:
Good catch; done in 35bbe82dc22d34c5db811e76d25233b7e4958afa

JelleZijlstra commented on stdlib/2/urllib2.pyi:
```diff
@@ -22,7 +33,7 @@ class Request(object):
     type = ...  # type: Optional[str]
     origin_req_host = ...
-    unredirected_hdrs = ...
+    unredirected_hdrs: Dict[str, str] = ...
```
Can remove the `= ...`. (I'm guessing this also supports unicode keys and values, but typing that correctly is hard.)

pmhahn replied:
Done in 9fea4210f9b8d83812cf4c54ee8655c4a64c264f

JelleZijlstra commented on stdlib/2/urllib2.pyi:
```diff
@@ -45,14 +56,21 @@ class Request(object):
+    def getcode(self) -> Optional[int]: ...
+
 class OpenerDirector(object):
+    addheaders: List[Tuple[str, str]] = ...
```
Remove the `= ...`.

pmhahn replied:
Done in b994fd5d89edf5bd7b3a2d3dee76b0529209f24e

JelleZijlstra commented on stdlib/2/urllib2.pyi:
```diff
@@ -45,14 +52,21 @@ class Request(object):
     def get_header(self, header_name: str, default: Optional[str] = ...) -> str: ...
     def header_items(self): ...
 
+class _Response(_SupportsRead, metaclass=ABCMeta):
```
Would this need to have more methods now that were removed from `_SupportsRead`? What methods is this object documented to support?

srittau commented on stdlib/2/urllib2.pyi:
```diff
@@ -68,50 +82,53 @@ class HTTPErrorProcessor(BaseHandler):
-    def add_password(self, realm: _string, uri: Union[_string, Sequence[_string]], user: _string, passwd: _string) -> None: ...
-    def find_user_password(self, realm: _string, authuri: _string) -> Tuple[Any, Any]: ...
+    def add_password(self, realm: Optional[_string], uri: Union[_string, Sequence[_string]], user: _string, passwd: _string) -> None: ...
+    def find_user_password(self, realm: Optional[_string], authuri: _string) -> Tuple[Any, Any]: ...
```
`Text` parameters should accept both `str` and `unicode` per PEP 484. (It's a bit hidden in the Python 2.7 section.) `unicode` won't work, since type checkers are not required to accept `str` arguments. mypy accepts both (although that might change), pytype does not.

pmhahn replied:
Fixed in f6a29d04fea7318e9bd7f94ce05c4c4afc45ae21
I only changed the type for the password handling, `_string` is still used by all the other methods.

pmhahn commented:
I rebased the whole branch due to one more merge conflict with master. I squashed all incremental commits and changed the code to use `addinfourl` instead of my own protocol `_SupportsRead` as inspired by 1442cc02.
